### PR TITLE
Updates to Armeria 0.87 and addresses deprecations

### PR DESCRIPTION
**Pull request by codefromthecrypt**

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
@anuraaga @trustin This update fails with a message like this.  It fails when gzip is used, either `ITZipkinHttpCollector.gzipEncoded()` or running the server manually and sending a POST with gzip to it.

Here's an example from running the server directly
```
127.0.0.1:50106][h1c] Unexpected exception:

io.netty.channel.socket.ChannelOutputShutdownException: Channel output shutdown
	at io.netty.channel.AbstractChannel$AbstractUnsafe.shutdownOutput(AbstractChannel.java:646) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.flush0(AbstractChannel.java:954) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.flush0(AbstractNioChannel.java:360) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.flush(AbstractChannel.java:906) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.flush(DefaultChannelPipeline.java:1370) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:739) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:731) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:717) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.ChannelDuplexHandler.flush(ChannelDuplexHandler.java:117) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:739) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:731) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:717) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.handler.flush.FlushConsolidationHandler.flushNow(FlushConsolidationHandler.java:206) [netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.handler.flush.FlushConsolidationHandler.flushIfNeeded(FlushConsolidationHandler.java:199) [netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.handler.flush.FlushConsolidationHandler.resetReadAndFlushIfNeeded(FlushConsolidationHandler.java:194) [netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.handler.flush.FlushConsolidationHandler.channelReadComplete(FlushConsolidationHandler.java:147) [netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:390) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:377) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelReadComplete(AbstractChannelHandlerContext.java:370) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.handler.timeout.IdleStateHandler.channelReadComplete(IdleStateHandler.java:295) [netty-handler-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:390) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:377) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelReadComplete(AbstractChannelHandlerContext.java:370) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelReadComplete(DefaultChannelPipeline.java:1413) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:390) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:377) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelReadComplete(DefaultChannelPipeline.java:936) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:168) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905) [netty-common-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [netty-common-4.1.34.Final.jar!/:4.1.34.Final]
	at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: io.netty.util.IllegalReferenceCountException: refCnt: 0
	at io.netty.buffer.AbstractByteBuf.ensureAccessible(AbstractByteBuf.java:1441) ~[netty-buffer-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.buffer.AbstractByteBuf.checkIndex(AbstractByteBuf.java:1373) ~[netty-buffer-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.buffer.PooledUnsafeDirectByteBuf.internalNioBuffer(PooledUnsafeDirectByteBuf.java:331) ~[netty-buffer-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.ChannelOutboundBuffer.nioBuffers(ChannelOutboundBuffer.java:433) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.socket.nio.NioSocketChannel.doWrite(NioSocketChannel.java:389) ~[netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	at io.netty.channel.AbstractChannel$AbstractUnsafe.flush0(AbstractChannel.java:939) [netty-transport-4.1.34.Final.jar!/:4.1.34.Final]
	... 33 more
```

**Comment by codefromthecrypt** [comment]
The code involved in gzip isn't in this change, btw, it is here:

https://github.com/apache/incubator-zipkin/blob/master/zipkin-server/src/main/java/zipkin2/server/internal/ZipkinHttpCollector.java#L208-L230

**Comment by codefromthecrypt** [comment]
fyi for sanity I did build and test right before this commit to update, including running independently and using brave to send gzipped data to it.

File: zipkin-collector/scribe/src/main/java/zipkin2/collector/scribe/ScribeInboundHandler.java [hunk 1]
```diff
@@ -162,7 +162,7 @@ void maybeReadPayload(ChannelHandlerContext ctx) {
           buf.release();
         }
       } else {
-        returned.writeBytes(content.array(), content.offset(), content.length());
+        returned.writeBytes(content.array(), 0, content.length());
```
Comment (anuraaga) [inline zipkin-collector/scribe/src/main/java/zipkin2/collector/scribe/ScribeInboundHandler.java]: Just `writeBytes(content.array())`

Back to my original, somewhat broken version ;)

File: zipkin-server/src/main/java/zipkin2/server/internal/ZipkinHttpCollector.java [hunk 2]
```diff
@@ -136,7 +136,7 @@ HttpResponse validateAndStoreSpans(SpanBytesDecoder decoder, ServiceRequestConte
         } else {
           // Currently this will happen for gzip spans. Need to fix armeria's gzip decoder to allow
           // returning pooled buffers on request.
-          nioBuffer = ByteBuffer.wrap(content.array(), content.offset(), content.length());
+          nioBuffer = ByteBuffer.wrap(content.array(), 0, content.length());
```
Comment (anuraaga) [inline zipkin-server/src/main/java/zipkin2/server/internal/ZipkinHttpCollector.java]: Ditto

**Review by anuraaga — COMMENTED** [review]
Will take a look at the failure tomorrow

File: zipkin-collector/scribe/src/main/java/zipkin2/collector/scribe/ScribeInboundHandler.java (hunk 1 above)
Comment (codefromthecrypt) [reply zipkin-collector/scribe/src/main/java/zipkin2/collector/scribe/ScribeInboundHandler.java]: :)

File: pom.xml [hunk 3]
```diff
@@ -48,9 +48,9 @@
 
     <main.basedir>${project.basedir}</main.basedir>
 
-    <armeria.version>0.86.0</armeria.version>
+    <armeria.version>0.87.0</armeria.version>
     <!-- This is from armeria, but be careful to avoid >= v20 apis -->
-    <guava.version>27.1-jre</guava.version>
+    <guava.version>28.0-jre</guava.version>
```
Comment (anuraaga) [inline pom.xml]: Don't remember exactly what this is but I wonder if it's removable since we switched to `-protocol`. Armeria itself shades guava

Comment (codefromthecrypt) [reply pom.xml]: you might be right

**Comment by anuraaga** [comment]
We were releasing `msg.content()`, which isn't correct when the gzip converter replaces that with a different `content` so changed to make sure only the gzip converter owns the reference to `msg.content()`

**Comment by codefromthecrypt** [comment]
cool. on the guava downgrade, the most important thing would be checking zipkin-gcp still runs.. I think there's some dodging related to this

**Review by codefromthecrypt — COMMENTED** [review]
thanks for the help!

**Comment by codefromthecrypt** [comment]
will recut zipkin on green with this

**Comment by anuraaga** [comment]
To keep the variables low, reverted the Guava downgrade since not sure what version is actually correct (Elasticsearch started to fail) ;) Think the comment is out of date but not sure what it's supposed to be.

**Comment by codefromthecrypt** [comment]
rebased on master

**Pull request merged**